### PR TITLE
Use `path.join(__dirname, 'package.json')` to load version

### PR DESCRIPTION
**Pull request by jesec**
Otherwise, webpack loads the `package.json` of the user's application...

Fixes:

```
Trace: FastifyError [Error]: fastify-plugin: fastify-static - expected
'3.x' fastify version, '4.6.0'* is installed.

* version of user's application
```

Bug: #3110
Refs: #3113

Signed-off-by: Jesse Chan <jc@linux.com>

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by jesec** [comment]
cc: @mcollina @zekth

File: fastify.js [hunk 1]
```diff
@@ -3,9 +3,10 @@
 const Avvio = require('avvio')
 const http = require('http')
 const querystring = require('querystring')
+const path = require('path')
 let lightMyRequest
 
-const { version } = require('./package.json')
+const { version } = require(path.join(__dirname, './package.json'))
```
Comment (zekth) [inline fastify.js:9]:
```suggestion
const { version } = require(path.join(__dirname, 'package.json'))
```

**Review by zekth — CHANGES_REQUESTED** [review]
good catch but just a minor issue.

**Comment by zekth** [comment]
Just one question, is Webpack able to do this resolution with the loader? Saying that because we compute the path using `path.join` instead of litteral relative one. Have you tested this with webpack?

**Comment by jesec** [comment]
> Just one question, is Webpack able to do this resolution with the loader? Saying that because we compute the path using `path.join` instead of litteral relative one. Have you tested this with webpack?

Yes. With this change, Webpack loads fastify's `package.json` instead of my application's.

**Review by zekth — APPROVED** [review]

**Comment by zekth** [comment]
Sidenote: i think as we had couple of webpack related questions in several repos like : https://github.com/fastify/fast-json-stringify . It would be worth the shot to had a test stack for the webpack bundling part.

**Comment by merceyz** [comment]
How does webpack end up loading the users `package.json` file? This is relative require so webpack should already be doing this join for you

**Comment by jesec** [comment]
> How does webpack end up loading the users `package.json` file? This is relative require so webpack shouldn't climb up the file tree to find it.

Well. I have no idea about internals of Webpack. This "just happens".

**Comment by zekth** [comment]
I'd personnaly prefer we fix this issue : https://github.com/fastify/fastify/issues/3158 
in prior to do another change blindly for webpack packaging.

**Comment by jesec** [comment]
> I'd personnaly prefer we fix this issue : #3158
> in prior to do another change blindly for webpack packaging.

I think it is a good idea to have tests. 

But meanwhile, I think this bug should be fixed. 

In my case, the bug blocks migration to fastify.

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by mcollina** [comment]
Let's get this shipped

**Comment by jorrit** [comment]
When I build my project, I get the error `Critical dependency: the request of a dependency is an expression`. Why didn't `require('./package.json');` work?

**Comment by climba03003** [comment]
> 
> 
> When I build my project, I get the error `Critical dependency: the request of a dependency is an expression`. Why didn't `require('./package.json');` work?

Please provide a reproducible code snippets and raise a new issue.

**Comment by jorrit** [comment]
Done, see #3185.

**Comment by alexeychikk** [comment]
This is an issue for me. I bundle my Node.js server with `esbuild` and this dynamic require causes `Cannot find module` error, since it tries to find `fastify`'s package.json in my `dist` folder.
If it was `require('./package.json')` like before this "fix", everything would be correctly resolved and bundled at build time.

@jesec The problem is in your webpack configuration or the webpack itself. Such statement `require('./package.json')` resolves to a package.json file relative to the source file from which it was called. It could not resolve to your project's `package.json` unless your webpack configuration somehow messes with it.

**Comment by Eomm** [comment]
> This is an issue for me. I bundle my Node.js app with `esbuild` and this dynamic require causes `Cannot find module` error, since it tries to find `fastify`'s package.json in my `dist` folder.
> If it was `require('./package.json')` like before this "fix", everything would be correctly resolved and bundled at build time.
> 
> @jesec The problem is in your webpack configuration or the webpack itself. Such statement `require('./package.json')` resolves to a package.json file relative to the source file from which it was called. It could not resolve to your project's `package.json` unless your webpack configuration somehow messes with it.

It would be really helpful to create a new issue with the recap of all the uses case, such as

|case|esbuild|webpack|
|---|---|---|
require package.json| broke| works|
fs read file| works | borken

etc..

In this way, we could take a more coherent action

**Pull request merged**